File: 02-longest-subarray-sum-k/solution.py
# Longest Subarray with Sum K
# Time: O(n), Space: O(n)

from typing import List

class Solution:
    def longestSubarrayWithSumK(self, arr: List[int], k: int) -> int:
        prefix_sum_map = {}
        sum_val = 0
        max_len = 0
        
        for i, num in enumerate(arr):
            sum_val += num
            
            if sum_val == k:
                max_len = i + 1
            
            if sum_val - k in prefix_sum_map:
                max_len = max(max_len, i - prefix_sum_map[sum_val - k])
            
            if sum_val not in prefix_sum_map:
                prefix_sum_map[sum_val] = i
        
        return max_len

# Test
if __name__ == "__main__":
    sol = Solution()
    print(sol.longestSubarrayWithSumK([1, 2, 3, 1, 1, 1, 1], 3))  # Output: 3

File: 02-longest-subarray-sum-k/solution.js
// Longest Subarray with Sum K
// Time: O(n), Space: O(n)

class Solution {
    longestSubarrayWithSumK(arr, k) {
        const prefixSumMap = new Map();
        let sumVal = 0;
        let maxLen = 0;
        
        for (let i = 0; i < arr.length; i++) {
            sumVal += arr[i];
            
            if (sumVal === k) {
                maxLen = i + 1;
            }
            
            if (prefixSumMap.has(sumVal - k)) {
                maxLen = Math.max(maxLen, i - prefixSumMap.get(sumVal - k));
            }
            
            if (!prefixSumMap.has(sumVal)) {
                prefixSumMap.set(sumVal, i);
            }
        }
        
        return maxLen;
    }
}

// Test
if (require.main === module) {
    const sol = new Solution();
    console.log(sol.longestSubarrayWithSumK([1, 2, 3, 1, 1, 1, 1], 3));  // Output: 3
}

module.exports = Solution;
